[ENH] Enable 3-d support for gaussian high-pass

ambrosejcarr opened the pull request:
- Makes `gaussian_high_pass` and `gaussian_low_pass` consistent
- Fixes typing through use of `Number`
- Adds 3D support for gaussian_high_pass (`osmFISH` needs it)
- Adds a test for new functionality

ttung commented on starfish/pipeline/filter/gaussian_low_pass.py:
```diff
@@ -29,41 +32,44 @@ def __init__(self, sigma, is_volume: bool=False, verbose=False, **kwargs) -> Non
+    def add_arguments(cls, group_parser: argparse.ArgumentParser) -> None:
         group_parser.add_argument(
-            "--sigma", default=1, type=int, help="standard deviation of gaussian kernel")
+            "--sigma", default=1, type=float, help="standard deviation of gaussian kernel")
```
how does this work with volumes?

ttung replied:
Also, I don't see any CLI parameterization for is_volume

ambrosejcarr replied:
if a scalar is passed, it is broadcast across all axes of the volume, so valid inputs are a scalar or a tuple of length == len(array.shape)

ttung commented on starfish/pipeline/filter/gaussian_low_pass.py:
```diff
@@ -29,41 +32,44 @@ def __init__(self, sigma, is_volume: bool=False, verbose=False, **kwargs) -> Non
         -------
         np.ndarray :
-            Blurred data in same shape as input image.
+            Blurred data in same shape as input image, converted to np.uint16 dtype.
```
Should we constrain the input types? At the very least, I would imagine that the skimage limitations should apply here.

ambrosejcarr replied:
There's some craziness here. I'm just documenting it in this PR but I've added a todo to look into why this is here and hopefully remove it.

ambrosejcarr replied:
We can safely cast the image types, added a warning when that happens.

ttung commented on starfish/pipeline/filter/gaussian_high_pass.py:
```diff
@@ -29,33 +38,33 @@ def get_algorithm_name(cls) -> str:
-        over_flow_ind: np.ndarray[bool] = img < blurred
-        res: np.ndarray = img - blurred
-        res[over_flow_ind] = 0
+        over_flow_ind: np.ndarray[bool] = image < blurred
```
I don't think this approach works for (-1, +1) ranged data.

ambrosejcarr replied:
You're right, it won't, however the gets cast to uint16 in `low_pass`, so that's not a problem here (although the _casting_ is potentially a big problem).

ambrosejcarr replied:
... and now that casting is explicit and won't break the results.

ttung commented on starfish/pipeline/filter/gaussian_high_pass.py:
```diff
@@ -68,9 +77,9 @@ def filter(self, stack: Stack) -> None:
 
-        high_pass: Callable = partial(self.gaussian_high_pass, sigma=self.sigma)
-        stack.image.apply(high_pass)
+        high_pass: Callable = partial(self.high_pass, sigma=self.sigma)
```
Idea: verify that if self.sigma is a tuple, that the length is equal to the shape of the image.

ambrosejcarr replied:
Done.

ambrosejcarr commented:
I've looked into the casting, which was definitely wrong. For the short term, a good solution is to warn the users that we only support uint16 image data here and properly cast to it using the `skimage` functionality. 

I've added some clipping to low_pass to prevent some crazy wrap-around behavior that was previously present, and verified that if passing a tuple, that it matches the expected image dimensions. 

Finally, I added CLI args.

ambrosejcarr commented on starfish/pipeline/filter/gaussian_high_pass.py:
```diff
@@ -1,26 +1,45 @@
+        sigma : Union[Number, Tuple[Number]]
             standard deviation of gaussian kernel
+        is_volume : bool
+            If True, 3d (z, y, x) volumes will be filtered, otherwise, filter 2d tiles independently.
```
This speaks to a general need to figure out what it means for 3d volumes to be first class. I suspect some approaches might do some weirds when presented with a (1, y, x).

I like `filter_by_2d_tile` as a flag with 3d as default but that could potentially be a lot of work to implement. I think the right thing here is to make an issue. What do you think?

ttung replied:
Cool that sounds good.

ttung replied:
I think the only downside of this approach is that the params/cmd line args would change, but I think we can live with that.

ambrosejcarr commented on starfish/pipeline/filter/gaussian_high_pass.py:
```diff
@@ -29,33 +48,40 @@ def get_algorithm_name(cls) -> str:
-        blurred: np.ndarray = GaussianLowPass.low_pass(img, sigma)
+        if image.dtype != np.uint16:
+            DataFormatWarning('gaussian filters currently only support uint16 images. Image data will be converted.')
+            image = img_as_uint(image)
```
Yea, `skimage` takes care of it so long as it's in a standard format (those two will be autodetected).

ttung replied:
I don't see how they can be autodetected.  There is no way to differentiate between an image that is entirely between midpoint grey and white, but in range (-1, +1), and an image that's (0, 1).

ambrosejcarr replied:
That's an interesting point, which provokes a question -- our image analysis is mostly about spot detection (bright against dark background). Do we think it matters that much that we can't differentiate these different types?

ambrosejcarr commented on starfish/pipeline/filter/gaussian_low_pass.py:
```diff
@@ -1,69 +1,92 @@
             image, sigma=sigma, output=None, cval=0, multichannel=True, preserve_range=True, truncate=4.0)
 
-        blurred = blurred.astype(numpy.uint16)
+        blurred = blurred.clip(0).astype(np.uint16)
```
The filter converts the data to float, and can produce values below zero because of how the gaussian convolution works. Those are not physiological and therefore not desirable, so we clip them.

ambrosejcarr replied:
I should also say: simple conversion to numpy.uint16 is a bug: the data wraps around. Very bad.

ttung replied:
Ohhh.

ambrosejcarr commented:
> I'm going to approve this because it's been pending forever, but I think we need to have a discussion about data types.

100% agree. I'm thinking a decorator that uses skimage to try to convert whatever users give us to uint16 might be the way to go. 

I'll rebase + merge this next time I have good internet.
